src/python/pants/core/util_rules/environments.py

```diff
@@ -377,6 +378,24 @@ class AllEnvironmentTargets(FrozenDict[str, Target]):
 class EnvironmentTarget:
     val: Target | None
 
+    def executable_search_path_cache_scope(
+        self, *, cache_failures: bool = False
+    ) -> ProcessCacheScope:
+        """Whether it's safe to cache exectuable search path discovery or not.
+
+        If the environment may change on us, e.g. the user upgrades a binary, then it's not safe to
+        cache the discovery to disk. Technically, in that case, we should recheck the binary every
+        session (i.e. Pants run), but we instead settle for every Pantsd lifetime to have more
+        acceptable performance.
+
+        Meanwhile, when running with Docker, we already invalidate whenever the image changes
+        thanks to https://github.com/pantsbuild/pants/pull/17101.
+        """
+        docker = self.val and self.val.has_field(DockerImageField)
+        if cache_failures:
+            return ProcessCacheScope.ALWAYS if docker else ProcessCacheScope.PER_RESTART_ALWAYS
+        return ProcessCacheScope.SUCCESSFUL if docker else ProcessCacheScope.PER_RESTART_SUCCESSFUL
```

Comment:
Is this something that could/should be done automatically as part of applying the `ProcessConfigFromEnvironment`?

Reply:
I don't think we can do that - note the `cache_failures` branch. That requires each call site tells us the behavior they want. `ProcessConfigFromEnvironment` does not allow us to have that per-call site change in behavior

Reply:
bump